prymer/offtarget/bwa.py

```diff
@@ -357,17 +360,17 @@ def _to_result(self, query: Query, rec: pysam.AlignedSegment) -> BwaResult:
                 "Query and Results are out of order" f"Query=${query.id}, Result=${rec.query_name}"
             )
 
-        num_hits: Optional[int] = int(rec.get_tag("HN")) if rec.has_tag("HN") else None
-        if rec.is_unmapped:
-            if num_hits is not None and num_hits > 0:
-                raise ValueError(f"Read was unmapped but num_hits > 0: {rec}")
-            return BwaResult(query=query, hit_count=0, hits=[])
-        elif num_hits > self.max_hits:
-            return BwaResult(query=query, hit_count=num_hits, hits=[])
-        else:
-            hits = self.to_hits(rec=rec)
-            hit_count = num_hits if len(hits) == 0 else len(hits)
-            return BwaResult(query=query, hit_count=hit_count, hits=hits)
+        num_hits: int = int(rec.get_tag("HN")) if rec.has_tag("HN") else 0
+        hits: list[BwaHit] = self.to_hits(rec=rec) if 0 < num_hits < self.max_hits else []
+
+        # `to_hits()` removes artifactual hits which span the boundary between concatenated
+        # reference sequences. If we are reporting a list of hits, the number of hits should match
+        # the size of this list. Otherwise, we either have zero hits, or more than the maximum
+        # number of hits. In both of the latter cases, we have to rely on the count reported in the
+        # `HN` tag.
+        num_hits = len(hits) if hits else num_hits
```

Comment:
```suggestion
        # `to_hits()` removes artifactual hits which span the boundary between concatenated
        # reference sequences. If we are reporting a list of hits, the number of hits should match
        # the size of this list. Otherwise, we either have zero hits, or more than the maximum
        # number of hits. In both of the latter cases, we have to rely on the count reported in the
        # `HN` tag.
        hits: list[BwaHit]
        if 0 < num_hits < self.max_hits:
            hits = self.to_hits(rec=rec)
            num_hits = len(hits)
        else:
            hits = []
```

Reply:
I think this probably handles edge cases better.  E.g. what if `HN=1` and that one hit spans a boundary, meaning you get back `[]` from `self.to_hits()`

Reply:
Had the same thought this morning!